feat(rx72n): Implement HARQ, FEC, and Hardware CRC-32 Protocol Stack

CesMag opened the pull request:
## Overview
This PR implements the robust **Hybrid ARQ (HARQ) Type I** communication stack for the RX72N firmware, fulfilling the requirements of **ADR-001 (ARQ to HARQ Transition)**. It replaces the simple ARQ mechanism with a reliable, high-speed protocol stack capable of handling the 10 Mbps SPI link between the RPi5 and RX72N, even in noisy environments.

## Key Features

### 1. Forward Error Correction (FEC)
- **Algorithm:** NASA-standard **K=7, Rate 1/2 Convolutional Code**.
- **Decoder:** Soft-decision **Viterbi Decoder** for maximum likelihood estimation.
- **Optimization:** Static memory allocation (no `malloc`) for deterministic behavior and safety.

### 2. Hybrid ARQ (HARQ)
- **Type:** HARQ Type I with **Chase Combining**.
- **Mechanism:** Accumulates soft bits from failed transmission attempts to improve Signal-to-Noise Ratio (SNR) before decoding.
- **Reliability:** Guarantees delivery with configurable retries and automatic combining.

### 3. Framing & Integrity
- **Framing:** Robust frame structure with **SYNC word**, Sequence Numbers, and Flags.
- **Integrity:** **IEEE 802.3 CRC-32** checksum for reliable error detection.
- **Acceleration:** Uses **RX72N Hardware CRC Calculator** peripheral for zero-overhead checksumming on target, with a bit-exact software fallback for host-based testing.

### 4. Protocol Buffers Integration
- **Serialization:** Integrated **nanopb** for efficient, type-safe serialization of motor commands and telemetry.
- **Messages:** Implemented handlers for `SetVelocityRequest`, `EmergencyStopRequest`, and telemetry.

### 5. SPI Communication Task
- **Architecture:** New **ThreadX task** (`spi_comm_task`) polling at 100Hz.
- **Flow:** Handles full command lifecycle: Receive \(\rightarrow\) Decode (FEC/CRC) \(\rightarrow\) Dispatch \(\rightarrow\) Execute \(\rightarrow\) Reply.

## Verification & Testing
- **Unit Tests:** Comprehensive host-based unit tests added in `star-rx72n-firmware/tests/`.
- **Coverage:**
  - ✅ **CRC-32:** Verified bit-exact compatibility with IEEE 802.3 and Go standard library.
  - ✅ **FEC:** Verified encoding and error-correcting Viterbi decoding.
  - ✅ **HARQ:** Verified Chase Combining logic and retry state machine.
  - ✅ **Framing:** Verified encoding/decoding round-trips and boundary checks.

## Hardware Details
- **CRC Hardware:** Configured Renesas RX `CRCCR` for IEEE 802.3 polynomial (`0x04C11DB7`) in LSB-first (reflected) mode.
- **Safety:** Strict adherence to **Power of 10** rules (no dynamic allocation, static analysis friendly).

## Related Issues
- Implements ADR-001
- Foundation for reliable RPi5 communication

CesMag committed:
Add unit tests for RX frame and HARQ protocol

- Implemented unit tests for the RX frame layer, including encoding, decoding, and utility functions.
- Added tests for encoder and decoder initialization, error handling, and round-trip encoding/decoding.
- Created a new test file for the HARQ protocol, covering initialization, state management, encoding, decoding, and combining functionality.
- Verified compatibility with the Go implementation through comprehensive test cases.

CesMag committed:
Add hardware and software CRC-32 implementations with comprehensive tests

- Introduced a new CRC calculator in `rx72n_regs.h` for hardware acceleration of CRC-32.
- Implemented hardware CRC-32 functions in `rx_crc32_hw.c`, utilizing the RX72N CRC peripheral.
- Developed a software fallback CRC-32 implementation in `rx_crc32_sw.c` for host testing and incremental updates.
- Created an internal API in `rx_crc_internal.h` to manage CRC initialization, deinitialization, and calculations.
- Added unit tests in `test_rx_crc32.c` to validate CRC functionality, including initialization, large buffer processing, unaligned data handling, and edge cases for null and zero-length inputs.
- Updated CMake configuration to include both hardware and software CRC sources, ensuring compatibility for host testing.

CesMag committed:
Add CRC-32 implementation details and protocol stack libraries to documentation

CesMag committed:
Add motor index enumeration and update velocity command handling

bsikar commented on star-rx72n-firmware/include/tasks/spi_comm_task.h:
```diff
@@ -0,0 +1,114 @@
+ * =============================================================================
+ */
+
+/**
```
I see that this section uses 4 spaces not 2, I ran the script to format the code and 31 files were changed. You did not run the formatting script before making this MR. Run that script.

```
/**
 * @brief SPI communication task configuration
 */
typedef enum {
    k_spi_comm_task_poll_hz     = 100, /**< SPI polling frequency (Hz) */
    k_spi_comm_task_period_ms   = 10,  /**< Polling period (ms) - 1000/100 */
    k_spi_comm_task_stack_size  = 4096, /**< Stack size (bytes) */
    k_spi_comm_task_priority    = 8,    /**< Task priority (lower than motor) */
    k_spi_comm_rspi_channel     = 0,    /**< RSPI channel for RPi5 */
    k_spi_comm_spi_mode         = 0,    /**< SPI mode 0 (CPOL=0, CPHA=0) */
} spi_comm_task_config_t;
```

IMMZEK replied:
fuck you

bsikar commented on star-rx72n-firmware/lib/rx_fec/inc/rx_fec.h:
```diff
@@ -0,0 +1,265 @@
+ * @brief Forward Error Correction (FEC) Codec for RX72N
+ *
+ * Implements NASA-standard K=7 convolutional encoder and Viterbi decoder.
+ * This is a C port of the Go FEC implementation in star-gateway for
```
Why are we using Viterbi, is `BCJR` not better? I know that BCJR requires more computation power, but we have a good chip, and we aren't utilizing much of its processing power in this application. The tradeoff might be worth it for us–I know that viterbi is often "good enough," but I think we should use BCJR if we can.

bsikar commented on star-rx72n-firmware/lib/rx_fec/inc/rx_fec.h:
```diff
@@ -0,0 +1,265 @@
+{
+    return (soft >= 0) ? 1 : 0;
+}
+
```
We should not have static methods inside header files, if there is a good reason for this please justify it with a comment in Github and also leave a comment here. When we use static methods inside headers it results in a larger binary and often its not needed. Usually, we should be placing the static methods inside the source `.c` file instead. Also refer to the style guide for understanding the prefix `internal_` and `priv_`. If this is meant to be exposed and not actually static use `priv_` but if its supposed to be static for only this `.c` then use `internal_` and place this static method inside the source.

bsikar commented on star-rx72n-firmware/lib/rx_fec/src/rx_fec.c:
```diff
@@ -0,0 +1,461 @@
+    size_t bit_pos  = 7 - (bit_idx % 8); /* MSB first */
+    return (data[byte_idx] >> bit_pos) & 1;
+}
+
```
Reference the previous comment about size_t

bsikar commented on star-rx72n-firmware/lib/rx_fec/src/rx_fec.c:
```diff
@@ -0,0 +1,461 @@
+static void internal_encode_bit(uint8_t *state, uint8_t input_bit,
+                                uint8_t *out0, uint8_t *out1)
+{
+    /* Shift in the new bit (input is MSB of the combined state) */
```
Since we are modifing the pointer a comment should be left in the doxygen to express that.

bsikar commented on star-rx72n-firmware/lib/rx_fec/src/rx_fec.c:
```diff
@@ -0,0 +1,461 @@
+    /* Check if we have enough stack space - use static buffer for safety */
+    /* For production, this should be a caller-provided buffer */
+    static rx_soft_bit_t soft_bits_buffer[RX_FEC_MAX_SYMBOLS * 2];
+
```
This variable is static, express why. Also the comment "For production", yeah this code should be setup for prod. This is bad. And verify would this lead to an off by 1 error, should it be `* 2` or `* 2 + 1`.

bsikar commented on star-rx72n-firmware/lib/rx_frame/inc/rx_crc_internal.h:
```diff
@@ -0,0 +1,119 @@
+#elif !defined(RX_CRC32_USE_SOFTWARE)
+#define RX_CRC32_USE_SOFTWARE
+#endif
+
```
We should only support hardware not software

bsikar commented on star-rx72n-firmware/lib/rx_frame/inc/rx_crc_internal.h:
```diff
@@ -0,0 +1,119 @@
+ * @return CRC-32 checksum (0 if data is NULL or len is 0)
+ */
+uint32_t rx_crc32_ieee_impl(const uint8_t *data, size_t len);
+
```
dont use size_t

bsikar commented on star-rx72n-firmware/lib/rx_frame/inc/rx_frame.h:
```diff
@@ -0,0 +1,279 @@
+{
+    return (type >= k_frame_type_command && type <= k_frame_type_nack);
+}
+
```
refer to the previous comment about static inside header

bsikar commented on star-rx72n-firmware/lib/rx_frame/src/rx_frame.c:
```diff
@@ -0,0 +1,284 @@
+    buf[1] = (uint8_t)((val >> 8) & 0xFF);
+    buf[2] = (uint8_t)((val >> 16) & 0xFF);
+    buf[3] = (uint8_t)((val >> 24) & 0xFF);
+}
```
we shouldnt be indexing with numbers, this should be an enum so we understand what each index represents

bsikar commented on star-rx72n-firmware/lib/rx_frame/src/rx_frame.c:
```diff
@@ -0,0 +1,284 @@
+/**
+ * @file rx_frame.c
```
I dont like how this has magic number that are not documented, `8`, `24`, etc seem magic same with `0xFF` we need a comment or this should be in an enum or a macro if needed so we can know why we are shifting like this

bsikar commented on star-rx72n-firmware/lib/rx_harq/src/rx_harq.c:
```diff
@@ -0,0 +1,378 @@
+                        size_t soft_len, size_t expected_output_len,
+                        uint8_t *output, size_t *output_len)
+{
+    if (harq == NULL || soft_bits == NULL || output == NULL ||
```
Method is too long

bsikar commented on star-rx72n-firmware/lib/rx_harq/src/rx_harq.c:
```diff
@@ -0,0 +1,378 @@
+
+    /* Get combined soft bits */
+    static rx_soft_bit_t combined_buffer[RX_HARQ_SOFT_BUFFER_SIZE];
+    size_t combined_len;
```
Why is this variable static

bsikar reviewed (changes requested):
I left lots of comments, here are some comments that I did not leave, but you also must do:

---
  CRITICAL ISSUES - Must Fix Before Merge

  Comment 1: Thread Safety Issue in HARQ Decoder

  File: star-rx72n-firmware/lib/rx_harq/src/rx_harq.c
  Line: 300

  **Thread Safety Violation**

  This static buffer creates a race condition if the HARQ handle can be used from multiple threads:

  ```c
  static rx_soft_bit_t combined_buffer[RX_HARQ_SOFT_BUFFER_SIZE];

  Problem: If rx_harq_decode() is called concurrently from different contexts (e.g., multiple SPI handles), they will corrupt each other's data.

  Solutions (choose one):
  1. Move this buffer into rx_harq_handle_t structure (preferred - makes handle truly independent)
  2. Document that each thread MUST have its own HARQ handle
  3. Add a mutex to protect this function (adds overhead)

  Recommendation: Move to handle structure for true thread safety.

  ---

  ### Comment 2: Message Type Encoding Not Documented

  **File:** `star-rx72n-firmware/src/tasks/spi_comm_task.c`
  **Line:** 246-248

  Undocumented Protocol Decision

  This uses a byte prefix to identify message types, but this isn't documented anywhere:

  uint8_t msg_type = frame->payload[0];
  const uint8_t *payload_data = &frame->payload[1];

  Issues:
  1. Not documented in docs/sections/01_nanopb_protocol.tex
  2. Not verified against Go gateway implementation
  3. Wastes a byte per message
  4. Could use the frame type field instead

  Required Actions:
  1. Add documentation to 01_nanopb_protocol.tex explaining this encoding
  2. Verify Go gateway expects this format
  3. Consider alternative: use frame.header.type field directly

  Alternative approach:
  // Use frame type directly instead of payload prefix
  switch (frame->header.type) {
      case k_frame_type_velocity_cmd:
          err = internal_handle_velocity_command(frame->payload, frame->header.length);
          ...
  }

  ---

  ### Comment 3: FEC Disabled by Default

  **File:** `star-rx72n-firmware/src/tasks/spi_comm_task.c`
  **Line:** 389

  FEC/HARQ Disabled Despite Full Implementation

  .fec_enabled = 0, /* FEC disabled for now */

  Concern: You've implemented comprehensive FEC/HARQ but it's disabled. This suggests:
  - Not fully tested in integration?
  - Performance concerns?
  - Waiting for RPi5 implementation?

  Required Action: Please add:
  1. A TODO with specific timeline or trigger for enabling FEC
  2. OR explanation of why it will remain disabled
  3. OR enable it and test

  Suggestion:
  .fec_enabled = 0, /* TODO(#XX): Enable after RPi5 gateway FEC implementation complete */

  ---

  ### Comment 4: Incomplete Telemetry Handler

  **File:** `star-rx72n-firmware/src/tasks/spi_comm_task.c`
  **Line:** 268-272

  Stub Implementation

  case k_msg_type_telemetry:
      /* Telemetry request - respond with current status */
      RX_LOG_DEBUG(s_tag, "Telemetry request received");
      /* TODO: Implement telemetry response */
      break;

  Decision needed:
  - If telemetry will be implemented soon, add specific TODO with issue number
  - If not needed yet, remove this case entirely to avoid confusion
  - Silently accepting but not responding could confuse the RPi5

  Recommendation: Send a response indicating "not implemented" or remove the case.

  ---

  ### Comment 5: Unknown Message Type - Missing NACK

  **File:** `star-rx72n-firmware/src/tasks/spi_comm_task.c`
  **Line:** 274-277

  Protocol Error Not Signaled to Controller

  default:
      RX_LOG_WARN(s_tag, "Unknown message type");
      err = RX_ERR_PROTOCOL_ERROR;
      break;

  Issue: The RX72N logs the error but doesn't inform the RPi5.

  Fix: Send NACK so the controller knows the command was rejected:
  default:
      RX_LOG_WARN(s_tag, "Unknown message type: 0x%02X", msg_type);
      err = RX_ERR_PROTOCOL_ERROR;
      // Inform controller of protocol error
      rx_spi_comm_send_nack(&s_spi_handle, frame->header.sequence,
                            k_frame_flag_protocol_error);
      break;

  ---

  ## **MINOR STYLE VIOLATIONS - Must Fix**

  ### Comment 6: Magic Number - CRC Final XOR

  **File:** `star-rx72n-firmware/lib/rx_frame/src/rx_crc32_hw.c`
  **Line:** 126

  Style Guide Violation: Magic Number

  Per docs/sections/05_c_style_guide.tex Section 3.3.2: "Constants and macros use SCREAMING_SNAKE_CASE"

  return CRC.CRCDOR.LONG ^ 0xFFFFFFFF;

  Fix:
  #define CRC32_IEEE_FINAL_XOR (0xFFFFFFFF)
  return CRC.CRCDOR.LONG ^ CRC32_IEEE_FINAL_XOR;

  Also applies to:
  - rx_crc32_sw.c:104, 113, 146, 155 (4 more instances)

  ---

  ### Comment 7: Magic Number - CRC Init Value

  **File:** `star-rx72n-firmware/lib/rx_frame/src/rx_crc32_sw.c`
  **Line:** 146

  Style Guide Violation: Magic Number

  uint32_t crc = 0xFFFFFFFF;

  Fix:
  #define CRC32_IEEE_INIT_VALUE (0xFFFFFFFF)
  uint32_t crc = CRC32_IEEE_INIT_VALUE;

  This constant should be defined once in a header (e.g., rx_crc_internal.h) and used everywhere.

  ---

  ### Comment 8: Magic Number - FEC Branch Metric Offset

  **File:** `star-rx72n-firmware/lib/rx_fec/src/rx_fec.c`
  **Line:** 141

  Style Guide Violation: Magic Number

  return 32768 - correlation;

  Fix:
  #define FEC_BRANCH_METRIC_OFFSET (32768)  /* Offset to keep metrics positive */

  // In function:
  return FEC_BRANCH_METRIC_OFFSET - correlation;

  The comment in line 121 is good, but the constant should be named.

  ---

  ### Comment 9: Magic Number - Busy-Wait Loop Count

  **File:** `star-rx72n-firmware/lib/rx_spi_comm/src/rx_spi_comm.c`
  **Line:** 304

  Style Guide Violation: Magic Number

  for (volatile uint32_t i = 0; i < 10000; i++) {
      __asm__ volatile("nop");
  }

  Fix:
  #define SPI_BUSY_WAIT_NOP_COUNT (10000)  /* Approximate 1ms delay at 240MHz */

  for (volatile uint32_t i = 0; i < SPI_BUSY_WAIT_NOP_COUNT; i++) {
      __asm__ volatile("nop");
  }

  Note: The TODO above this code mentions replacing with ThreadX semaphore, which is the proper long-term solution. This is just for short-term clarity.

IMMZEK commented on docs/sections/01_nanopb_protocol.tex:
```diff
@@ -163,6 +163,57 @@ \subsubsection{Byte Order and Standards Compliance}
+uint32_t rx_crc32_update(uint32_t crc, const uint8_t *data, size_t len);
+\end{lstlisting}
+
+\textbf{Go Compatibility:} Both implementations produce bit-exact results matching Go's \texttt{crc32.ChecksumIEEE()}, verified by 18 unit tests including the canonical test vector ``123456789'' = \texttt{0xCBF43926}.
```
Done

IMMZEK commented on star-rx72n-firmware/lib/rx_fec/src/rx_fec.c:
```diff
@@ -0,0 +1,461 @@
+ */
+static void internal_init_branch_table(
+    uint8_t branch_table[k_fec_num_states][2][2])
+{
```
i likey magic number

IMMZEK commented on star-rx72n-firmware/lib/rx_hal/inc/rx72n_regs.h:
```diff
@@ -379,6 +379,57 @@ typedef struct {
+    volatile uint16_t WORD; /* 16-bit access for CRC-16 */
+    volatile uint8_t  BYTE; /* 8-bit access for CRC-8 */
+  } CRCDOR;                 /* 0x08 - CRC Data Output Register */
+} CRC_Type;
```
good job we didnt need the other stuff

CesMag committed:
fixed brightons bullshit

CesMag committed:
Exclude nanopb submodule from format script

bsikar commented:
It all looks good, but we are still using size_t in places and we shouldnt be. Use uint32_t instead

IMMZEK commented:
Done with `size_t`

CesMag committed:
Refactor size_t to uint32_t in FEC, HARQ, and Nanopb modules

- Updated function signatures and internal implementations to replace `size_t` with `uint32_t` for consistency and to avoid potential issues with size mismatches.
- Modified related test cases to accommodate the new data type.
- Ensured that all relevant files in the rx_fec, rx_harq, and rx_nanopb libraries reflect these changes.
- Adjusted documentation and comments where necessary to maintain clarity.
